fix(bazel): do not use manifest paths for generated imports within compilation unit

devversion opened the pull request:
Currently, the `ng_module` rule incorrectly uses manifest paths for
generated imports from the Angular compiler.

This breaks packaging as prodmode output (i.e. `esnext`) is copied in
various targets (`es5` and `es2015`) to the npm package output. Example:

_node_modules/my-pkg/es2015/imports/public-api.js_
```ts
import * as i1 from "angular/packages/bazel/test/ng_package/example/imports/second";
```

The above import is invalid in NPM and leaks information about the source project
structure. The import should be relative:

```ts
import * as i1 from "./second";
```

The imports can, and **should** be relative so that the files of the same compilation unit are self-contained and do not rely on custom module resolution.

devversion committed:
test: add entry-point with generated imports to ng_package test

gregmagolan reviewed (approved):
LGTM. This code is much too complicated. We will be able to replace this with ts_library with angular as a plugin in the future once we can drop
ViewEngine support.

devversion commented:
Yep. Totally agreed. And yes! I'm waiting for that to happen. The current system is too complicated. I tried to improve this by adding comments. Also the fact that we rely on summaries, but not on metadata files, makes ngc-wrapped even more complicated.

kyliau commented:
g3 presubmit: http://test/OCL:298679132:BASE:298679262:1583270817811:1d706ca6

kyliau commented:
g3 presubmit: http://test/OCL:298679132:BASE:298853465:1583338254767:e12dceb4

kyliau commented:
g3 presubmit: http://test/OCL:298679132:BASE:298918004:1583355304817:c6251c4b

devversion committed:
fix(bazel): do not use manifest paths for generated imports within compilation unit

Currently, the `ng_module` rule incorrectly uses manifest paths for
generated imports from the Angular compiler.

This breaks packaging as prodmode output (i.e. `esnext`) is copied in
various targets (`es5` and `es2015`) to the npm package output.

e.g. imports are generated like:

_node_modules/my-pkg/es2015/imports/public-api.js_
```ts
import * as i1 from "angular/packages/bazel/test/ng_package/example/imports/second";
```

while it should be actually:

```ts
import * as i1 from "./second";
```

The imports can, and should be relative so that the files are
self-contained and do not rely on custom module resolution.

kyliau commented:
g3 presubmit: http://test/OCL:298679132:BASE:299157401:1583438236266:2d80d68f

kyliau commented on packages/bazel/src/ngc-wrapped/index.ts:
```diff
@@ -290,20 +291,32 @@ export function compile({allDepsCompiledWithBazel = true, compilerOpts, tsHost,
 
-  function doFileNameToModuleName(importedFilePath: string): string {
+  function doFileNameToModuleName(importedFilePath: string, containingFilePath?: string): string {
+    const relativeTargetPath =
```
this needs to be moved to line 358 to preserve the original behavior.

kyliau commented:
g3 presubmit: http://test/OCL:298679132:BASE:299217178:1583452908596:d0fe404c
